### ISSUE TITLE
Unicode UTF8 dont displayed correctly

### ISSUE THREAD

**Issue by JaydenSu** [raysan5/raylib#2991]
In continuation of the issue: https://github.com/raysan5/raylib/issues/2022

If I understood correctly from this "issue", then the code:

```
std::string string_ = u8"Comença a jugar çà";
DrawText(&string_[0], rec.x + 10, rec.y + 60, 60, BLACK);
```

Should be displayed on the raylib window without any problems - and it is, everything is displayed perfectly.

BUT, this code does not want to be displayed correctly:

```
string_ = u8"Click me!东东2";
DrawText(&string_[0], rec.x + 10, rec.y + 10, 60, BLACK);
```

Please tell me - why?

![image](https://user-images.githubusercontent.com/128853845/227798047-af6629a9-3122-4831-be56-46c9e9f80f03.png)

**Comment by raysan5** [comment]
You need to provide a font containing the required glyphs and load them. Please, check the `text_unicode` example.

Please, also note that this is not a forum for questions, follow the required Issues template before publishing new issues or ask in raylib Discord, probably the community can hel with this kind of questions.

**Comment by JaydenSu** [comment]
Thank you.

By the way, the example from "text_unicode.c" - doesn't work if I pass a string:
u8"东东东"

PS: That is, in fact, we can say in raylib There is no support for Unicode and UTF8 encoding in particular.

**Comment by orcmid** [comment]
You cannot say that.  It depends on the availability of the fonts.  The UTF8 encodings should be fine.

If you display those strings in hexadecimal, they should reveal that correct UTF8 is employed in both cases.

raylib is not responsible for the support of CJK glyphs in a chosen font.  Font providers handle that or not.  I suspect the default font used in raylib DrawText() does not support any glyphs except those graphical ones in u0000 - u00ff, and that's what you have demonstrated.

As Ramon suggests, ask for font solutions on Discord, although I would expect that your computer has one, because you can enter and see them.

**Comment by JaydenSu** [comment]
> You cannot say that. It depends on the availability of the fonts. The UTF8 encodings should be fine.
I can, I already said.

> As Ramon suggests, ask for font solutions on Discord, although I would expect that your computer has one, because you can enter and see them.

The fonts that are in the examples work.
But firstly, they do not work by default.
And secondly - they work, my changes, but they work through the ass.

The fonts that are in the examples work.
But firstly, they do not work by default.
And secondly - they work, my changes, but they work through the ass.

Therefore, we can say that raylib does not, in fact, support the output of Unicode text, or it does, but with very complex manipulations.

That is, you cannot receive, for example, an array of bytes encoded for example UTF8 from the server and display it in the window. And this is a fact.

**Comment by raysan5** [comment]
I'm sorry it didn't worked for you. Please note raylib is a free and open source project, you can contribute with features, improvements or clearer examples. Contributions are welcome.

**Comment by orcmid** [comment]
@JaydenSu 

I can see the calligraphy in

u8"东东东"

because I do have a font to do it (or GitHub is creating tiny images, which I do not believe is the case).  That does not mean that everyone on GitHub can see it.

In email, the situation is worse.  There are many characters I cannot use because different readers (or the list servers that forward to them) do not have the fonts or do not recognize/preserve messages in UTF-8.

That I have the fonts on my machine already is happenstance, although perhaps likely these days.  

There are other situations such as running a console application, where it is more difficult to get Unicode rendering even though I know the application is using UTF8.

raylib default fonts in examples is another such case.  What raylib does not do is selective font fall-back and substitution when a Unicode code point is not satisfied in the current font and there are other (known) fonts that have glyphs for that code point.  That is much to expect from raylib.  Also from the underlying graphical libraries that are used for text display.  

So would you prefer that it be declared raylib does not handle UTF8 at all?  Then we'd have to select code pages and double-byte encoding schemes.  Is that an improvement?

With UTF8 recognized and handled, at least there is a direction for expanding the support of Unicode code points in text graphics at some point.  Are you prepared to dig into word-processing software, see how this is managed there, and find a comparable solution for raylib?

PS: I did paste u8"东东东" successfully into Microsoft Word, LibreOffice Writer, and Apache OpenOffice Writer documents.  The last two are open source and all three use an open format (OOXML and ODF) that can be inspected to see how that is accomplished in document parameters.

**Comment by JaydenSu** [comment]
> @JaydenSu
> 
> I can see the calligraphy in
> 
> u8"东东东"
> 
> because I do have a font to do it (or GitHub is creating tiny images, which I do not believe is the case). That does not mean that everyone on GitHub can see it.
> 
> In email, the situation is worse. There are many characters I cannot use because different readers (or the list servers that forward to them) do not have the fonts or do not recognize/preserve messages in UTF-8.
> 
> That I have the fonts on my machine already is happenstance, although perhaps likely these days.
> 
> There are other situations such as running a console application, where it is more difficult to get Unicode rendering even though I know the application is using UTF8.
> 
> raylib default fonts in examples is another such case. What raylib does not do is selective font fall-back and substitution when a Unicode code point is not satisfied in the current font and there are other (known) fonts that have glyphs for that code point. That is much to expect from raylib. Also from the underlying graphical libraries that are used for text display.
> 
> So would you prefer that it be declared raylib does not handle UTF8 at all? Then we'd have to select code pages and double-byte encoding schemes. Is that an improvement?
> 
> With UTF8 recognized and handled, at least there is a direction for expanding the support of Unicode code points in text graphics at some point. Are you prepared to dig into word-processing software, see how this is managed there, and find a comparable solution for raylib?
> 
> PS: I did paste u8"东东东" successfully into Microsoft Word, LibreOffice Writer, and Apache OpenOffice Writer documents. The last two are open source and all three use an open format (OOXML and ODF) that can be inspected to see how that is accomplished in document parameters.
> 
> PPS: You can compile UTF-8 into a program. How well are you able to output it in a shell/terminal application?

I understand it.
So, apparently, I'm too used to simply passing text in UTF16 encoding to WinApi functions, in particular GDI + functions, and they themselves figured out what and how to draw "under the hood".
Such functions can be called "easy to use", but in raillib - this is, unfortunately, Hell.

**Comment by orcmid** [comment]
@JaydenSu "So, apparently, I'm too used to simply passing text in UTF16 encoding to WinApi functions, in particular GDI + functions, and they themselves figured out what and how to draw 'under the hood'."

Yes, GDI is very high level in that respect, and there is an entire operating system behind it.  

Raylib is not at that level; it is much simpler in terms of what it makes easy to accomplish.  That's very different.  

I decided to look at something like DirectX to see what it affords.  This article is pretty scary compared to what raylib provides: https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/october/directx-factor-text-formatting-and-scrolling-with-directwrite

The interesting thing is at the beginninng.  A font was named as part of setting up text.  From the examples, it appears that font metrics are handled and so are a number of other features around types of fonts.  This all works by using platform-specific facilities.

I am reminded that raylib's inspiration is work by Borland in the 1980s.   None of these complexities and capabilities were prevalent in those times.

I am disappointed that your expectations are not satisfied.  It would be useful to explain and demonstrate the successful use of Unicode across a variety of languages and glyph forms.  That has not been done here.  It is probably best not to suggest much about the abiliity to pass UTF-8 (or Wide character UTF-16) around.

**Comment by chriscamacho** [comment]
https://en.wikipedia.org/wiki/2,147,483,647
Using a TTF font with the code points that you want to use is trivial and works, you just need to put some (minor) effort in.
Raylib is intended to be a LOW level C library, not an all in one high level library...

**Comment by orcmid** [comment]
@chriscamacho If those libraries were explicit about `uint32` rather than whatever the compiled-for model (signed) `int`, I would not have mentioned it.  Of course the largest used Unicode code point in UTF-8 is U+10FFFFF  as of Unicode 7.0.  I surmise the scheme can go to U+1FFFFFF without adding an additional first-byte rule making a 5-byte UTF-8 sequence.      

That's an incidental concern.  If modification of `rtext.c` for this purpose is trivial, it would be useful to have it done in raylib with or without any extension to the API.     My impression is that compiling with TTF enabled is not the default.   

It appears that the API has all the interfaces needed although I have some concern for what it would take to work with something like CJK with a ton of code points.  A code-point vector would be quite lengthy and then there is the generation of different font sizes for rasterizing. 

Is there advice about all this somewhere.  I am unfamiliar with the wiki.

**Comment by orcmid** [comment]
@JaydenSu  There are examples of non-trivial Unicode usage in the raylib/examples/text/ folder.  The *.png files there are informative.  The only use of CJK glyphs seems to be from a raster font, unfortunately (or the TTF-generated rasters are up-scaled).  See the text/resources/noto_cjk.png.

**Comment by designerfuzzi** [comment]
i can confirm it does not work. 
in my case 
```
int codepoints[] = {  0x0391, 0x0101};
int utf8Size = 2;
const char *text = CodepointToUTF8(codepoints[1], &utf8Size); 
```
utf8Size makes not even sense cause the standard is either 
`\U00001234`
or
`\u1234`
```
Font boldnumber = LoadFontEx(otfpath2, 48, codepoints, 1); //<-- crash
```
will crash.

Using the very same otf file in osx, running it thru font fontcheck and also using it in another app or even browser works just fine. The specific letter is a little larger than the lineheight. My guess is that the entire font loader forgot that letters can be larger than a simple rectangle of line-height and their kerning. It really is just a guess, or the glyph is not closed cause it is outside the boundary..

**Comment by orcmid** [comment]
@designerfuzzi 
> ```
> int codepoints[] = {  0x0391, 0x0101};
> int utf8Size = 2;
> const char *text = CodepointToUTF8(codepoints[1], &utf8Size);
> printf("%s",text); //prints just fine the letter.
> ```
> utf8Size makes not even sense ...

Please notice that `utf8Size` is given as an **output** parameter where the derived UTF8size is **returned**, not provided.

In your example, `codepoints[1]` of 0x0101 is for Unicode Latin Small Letter A with Macron (ā), and its UTF8 is the two-octet sequence `C4 81` so the UTF8 size for that one code point is indeed 2 bytes and that is what you should find in text[0] and text[1].

In the `LoadFontEx(otfpath2, 48, codepoints, 1);` you are loading the u+0391, not the u+0101.  Is that a problem there?  I don't see how that has anything to do with UTF8.

**Comment by designerfuzzi** [comment]
`&utf8Size` is an inout to be precise. Now that i read the `rtext.c` i know what it does. 

lets pinpoint whats happening.
`LoadFontEx(fontpath, 24, NULL, 0);` reads up to 95 glyphs as written in `rtext.c`.
The majority of unicodes are outside the range of `0..95` aka `0x00..0x5F`, which means any text that contains unicodes higher then \u005F (and that are basically all unicodes) only load when explicit codepoints are given, that makes perfectly sense only after reading rtext.c. 

As i know now **unicode-glyph loading works only when codepoints are known**, we give the codepoints table as argument upfront or parse the codepoints out of text with `LoadCodepoints()`. 

So found following works fine, 
```
const char *text = "\u0391\u0101"; //<-- official UTF-8 document unicode notation
int count = 0;
int * codepoints = LoadCodepoints(text, &count);
Font font = LoadFontEx(fontpath, 24, codepoints, count);
```
which was proof the font is not broken, the source has an undetected vulnerability

I first tried it with just two manually given codepoints without parsing from text as written in the example above. This way i crashed raylib with a freeze, not just error, the entire computer froze. So i went on seeking the error, and of course i did not exclude the possibility that fonts can be broken. But knowing it is not broken as i can perfectly load it in an ios app or mac app or web that scenario is highly unlikely. 

My earlier attempt with just two or one codepoint crashed at LoadFontEx() and then without report with a freeze. So i might have found something that is not excluded as scenario. 
Because if that is not closed, it is frankly spoken a vulnerability that can be exploited by simple text that can and will contain unicode letters. I was more expecting failure to result in `?` glyph or even `""` but it froze. 

crashing code:
```
int codepoints[] = {  0x0391, 0x0101};
Font font = LoadFontEx(fontpath, 24, codepoints,1);
```
So i tried
```
int codepoints[] = {  0x0391, 0x0101};
Font font = LoadFontEx(fontpath, 24, NULL,8000);
```
does not crash but eats memory of course.

And this tells me that there is a usecase that is not excluded.

**Comment by designerfuzzi** [comment]
feedback: found the problem.
in `rtext.c` function `GenImageFontAtlas()` 
the `fontSize` is taken as granted, therefore rendering in an insufficient image size can result in memory corruption. 

Means multiple functions did not detect the possibility that a font glyph can have an image height that does fit into the atlas, so packing fails on such glyphs, writing in memory that is not allocated likely corrupts the process. Which was the reason why i got mem checksum errors.

here my quick fix, for sure not ideal as it utilises fontSize.
`#define WORKAROUND`
```C
// Calculate image size based on total glyph width and glyph row count
int totalWidth = 0;
int maxGlyphWidth = 0;

WORKAROUND int maxGlyphHeight = 0;

for (int i = 0; i < glyphCount; i++)
{
    if (glyphs[i].image.width > maxGlyphWidth) maxGlyphWidth = glyphs[i].image.width;
    totalWidth += glyphs[i].image.width + 2*padding;
    WORKAROUND if (glyphs[i].image.height > maxGlyphHeight) maxGlyphHeight = glyphs[i].image.height;
}

WORKAROUND if (fontSize<maxGlyphHeight) fontSize = maxGlyphHeight;
```
Accordingly function `MeasureTextEx()` which is used when `ImageDrawTextEx()` is processed, also returns for sure only fontSize as evaluated height, no matter what, which is likely easy to code with but not correct - because the function makes one expect to get the actual size of text, rather then just evaluated width and non-evaluated ~~fontSize~~ height.

Means `MeasureTextEx()` could use an improvement as well.
like so .. perhaps.
```
if (letter != '\n')
{
    if (font.glyphs[index].advanceX != 0) textWidth += font.glyphs[index].advanceX;
    else textWidth += (font.recs[index].width + font.glyphs[index].offsetX);
    WORKAROUND if (textHeight < font.glyphs[index].image.height) textHeight = font.glyphs[index].image.height;
}
```

Last isse, 
handing out a `const char *` string solely with unicode can lead to missing nullTerminator depending on IDE. When i forced my `const char * text = "\u0391";` to have an nullTerminator no matter what, it never complains about its value. See where the problem comes from? Unicode can express control letters that trick debugger, precompiler in weirdo states. They must be nullterminated (actually one would expect they are by default via IDE). 
So i wrote a simplification for my case.. which looks like..
```
const char *nullTerminate(const char *inp) {
    size_t len = strnlen(inp, 255);
    char *txt = (char *)malloc(len+1);
    if (txt==NULL || inp==NULL) return "";
    txt[len] = '\0';
    while (len--) txt[len]=inp[len];
    return txt;
}

Font LoadFontWithCodepointsFromText(const char *path, int fontSize, const char *text) {
    const char *check = nullTerminate(text);
    int count = 0;
    int * codepoints = LoadCodepoints(check, &count);
    Font result = LoadFontEx(path, fontSize, codepoints, count);
    UnloadCodepoints(codepoints);
    return result;
}
```

Now everything works.. `MeasureTextEx()`, `ImageDrawTextEx()`, `DrawTextEx()` as expected. 

<img width="299" alt="Bildschirmfoto 2024-03-03 um 10 02 18" src="https://github.com/raysan5/raylib/assets/1221499/e98f0f80-24a2-4457-a99c-a137ade58e85">

Still visible that the TextureSize for the Atlas before packing is not ideal as it just takes the proportional attempt. But at least a starting point

**Comment by raysan5** [comment]
@designerfuzzi Wow! This is very interesting, I assumed the glyphs image height could not be bigger than requested font size... I'm opening a separate issue to review this: https://github.com/raysan5/raylib/issues/3858
